validation-test/stdlib/Zip.swift

```diff
@@ -0,0 +1,196 @@
+      stackTrace: SourceLocStack().with(test.loc))
+  }
+}
+
+ZipTests.test("Collections") {
+  typealias Element = (OpaqueValue<Int>, OpaqueValue<Int32>)
+  typealias Sequence = Zip2Sequence<
+    MinimalCollection<OpaqueValue<Int>>,MinimalCollection<OpaqueValue<Int32>>>
+  typealias Collection = Zip2Collection<
+    MinimalCollection<OpaqueValue<Int>>,MinimalCollection<OpaqueValue<Int32>>>
+
+  func compareElements(_ lhs: Element, rhs: Element) -> Bool {
+    return lhs.0.value == rhs.0.value && lhs.1.value == rhs.1.value
+  }
+
+  expectCollectionAssociatedTypes(
+    collectionType: Collection.self,
+    iteratorType: Sequence.Iterator.self,
+    subSequenceType: Slice<Collection>.self,
+    indexType: Collection.Index.self,
+    indicesType: DefaultIndices<Collection>.self)
+    
+  for test in zipTests {
+    let left = MinimalCollection<OpaqueValue<Int>>(
+      elements: test.sequence.map(OpaqueValue.init))
+    let right = MinimalCollection<OpaqueValue<Int32>>(
+      elements: test.other.map(OpaqueValue.init))
+
+    var result = zip(left, right)
+    expectType(Collection.self, &result)
+
+    checkCollection(
+      test.expected.map { (OpaqueValue($0), OpaqueValue($1)) }, 
+      result, sameValue: compareElements)      
+  }
+}
+
+ZipTests.test("RandomAccessCollections") {
+  typealias Element = (OpaqueValue<Int>, OpaqueValue<Int32>)
+  typealias Sequence = Zip2Sequence<
+    MinimalRandomAccessCollection<OpaqueValue<Int>>,
+    MinimalRandomAccessCollection<OpaqueValue<Int32>>>
+  typealias Collection = Zip2Collection<
+    MinimalRandomAccessCollection<OpaqueValue<Int>>,
+    MinimalRandomAccessCollection<OpaqueValue<Int32>>>
+
+  func compareElements(_ lhs: Element, rhs: Element) -> Bool {
+    return lhs.0.value == rhs.0.value && lhs.1.value == rhs.1.value
+  }
+      
+  expectCollectionAssociatedTypes(
+    collectionType: Collection.self,
+    iteratorType: Sequence.Iterator.self,
+    subSequenceType: Slice<Collection>.self,
+    indexType: Collection.Index.self,
+    indicesType: DefaultIndices<Collection>.self)
+
+  for test in zipTests {
+    let left = MinimalRandomAccessCollection<OpaqueValue<Int>>(
+      elements: test.sequence.map(OpaqueValue.init))
+    let right = MinimalRandomAccessCollection<OpaqueValue<Int32>>(
+      elements: test.other.map(OpaqueValue.init))
+
+    var result = zip(left, right)
+    expectType(Collection.self, &result)
+    checkRandomAccessCollection(
+      test.expected.map { (OpaqueValue($0), OpaqueValue($1)) }, 
+      result, sameValue: compareElements)
+  }
+}
+
+ZipTests.test("Collection.isEmpty") {
+  expectTrue(zip(0..<0,0..<0).isEmpty)
+  expectTrue(zip(0..<1,0..<0).isEmpty)
+  expectTrue(zip(0..<0,0..<1).isEmpty)
+  expectFalse(zip(0..<1,0..<1).isEmpty)
+  expectFalse(zip(0..<2,0..<3).isEmpty)
+  expectFalse(zip(0..<3,0..<2).isEmpty)
+}
+
+ZipTests.test("Collection.count") {
+  expectEqual(zip(0..<0,0..<0).underestimatedCount, 0)
```

Comment:
It turns out we never define `underestimatedCount` for `Zip2Sequence`. I think it should be `Swift.min(seq1.underestimatedCount, seq2.underestimatedCount)`.

Reply:
Also probably worth overriding `underestimatedCount` for `Zip2Collection` as well, returning the minimum of underestimated counts of the base collections, and only calling `count` for the `RandomAccessCollection` case.

Reply:
Yeah good point, will add that.